src/tests/mlkem.rs

```diff
@@ -0,0 +1,257 @@
+        mechanism: CKM_ML_KEM,
+        pParameter: std::ptr::null_mut(),
+        ulParameterLen: 0,
+    };
+    let mut ciphertext = [0u8; 800];
+    let mut outlen: CK_ULONG = 800;
+    let mut handle_enc = CK_INVALID_HANDLE;
+    let ret = fn_encapsulate_key(
+        session,
+        &mut mechanism,
+        pub_handle,
+        key_template.as_ptr() as *mut _,
+        key_template.len() as CK_ULONG,
+        ciphertext.as_mut_ptr(),
+        &mut outlen,
+        &mut handle_enc,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut handle_dec = CK_INVALID_HANDLE;
+    let ret = fn_decapsulate_key(
+        session,
+        &mut mechanism,
+        priv_handle,
+        key_template.as_ptr() as *mut _,
+        key_template.len() as CK_ULONG,
+        ciphertext.as_mut_ptr(),
+        outlen,
+        &mut handle_dec,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut value_enc = vec![0u8; 16];
+    let mut extract_template = make_ptrs_template(&[(
+        CKA_VALUE,
+        void_ptr!(value_enc.as_mut_ptr()),
+        value_enc.len(),
+    )]);
+
+    let ret = fn_get_attribute_value(
+        session,
+        handle_enc,
+        extract_template.as_mut_ptr(),
+        extract_template.len() as CK_ULONG,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut value_dec = vec![0u8; 16];
+    let mut extract_template = make_ptrs_template(&[(
+        CKA_VALUE,
+        void_ptr!(value_dec.as_mut_ptr()),
+        value_dec.len(),
+    )]);
+
+    let ret = fn_get_attribute_value(
+        session,
+        handle_dec,
+        extract_template.as_mut_ptr(),
+        extract_template.len() as CK_ULONG,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    assert_eq!(value_enc, value_dec);
+
+    /* try again with key generation first */
+    let (pub_handle, priv_handle) = ret_or_panic!(generate_key_pair(
+        session,
+        CKM_ML_KEM_KEY_PAIR_GEN,
+        &[
+            (CKA_CLASS, CKO_PUBLIC_KEY),
+            (CKA_KEY_TYPE, CKK_ML_KEM),
+            (CKA_PARAMETER_SET, CKP_ML_KEM_512)
+        ],
+        &[],
+        &[(CKA_ENCAPSULATE, true)],
+        &[
+            (CKA_CLASS, CKO_PRIVATE_KEY),
+            (CKA_KEY_TYPE, CKK_ML_KEM),
+            (CKA_PARAMETER_SET, CKP_ML_KEM_512)
+        ],
+        &[],
+        &[(CKA_DECAPSULATE, true),],
+    ));
+
+    let ret = fn_encapsulate_key(
+        session,
+        &mut mechanism,
+        pub_handle,
+        key_template.as_ptr() as *mut _,
+        key_template.len() as CK_ULONG,
+        ciphertext.as_mut_ptr(),
+        &mut outlen,
+        &mut handle_enc,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut handle_dec = CK_INVALID_HANDLE;
+    let ret = fn_decapsulate_key(
+        session,
+        &mut mechanism,
+        priv_handle,
+        key_template.as_ptr() as *mut _,
+        key_template.len() as CK_ULONG,
+        ciphertext.as_mut_ptr(),
+        outlen,
+        &mut handle_dec,
+    );
+    assert_eq!(ret, CKR_OK);
```

Comment:
Add also the extracting values + check they match?

Reply:
I have that mod locally already, will push soon.